src/client/datascience/kernel-launcher/kernelFinder.ts

```diff
@@ -0,0 +1,232 @@
+    ) {}
+
+    public async findKernelSpec(
+        interpreterUri: InterpreterUri,
+        token: CancellationToken,
+        kernelName?: string
+    ): Promise<IJupyterKernelSpec> {
+        const resource = isResource(interpreterUri) ? interpreterUri : undefined;
+        const notebookInterpreter = isResource(interpreterUri) ? undefined : interpreterUri;
+
+        if (kernelName) {
+            let kernelSpec = await this.findCachePath(kernelName);
+
+            if (kernelSpec) {
+                return kernelSpec;
+            }
+
+            if (!notebookInterpreter) {
+                kernelSpec = await this.getKernelSpecFromActiveInterpreter(resource, kernelName);
+            }
+
+            if (kernelSpec) {
+                return kernelSpec;
+            }
+
+            const kernelSearches = [
+                this.interpreterService.getInterpreters(resource).then((interpreters) => {
+                    const interpreterPaths = interpreters.map((interp) => interp.path);
+                    return this.findInterpreterPath(interpreterPaths, kernelName);
+                }),
+                this.findDiskPath(kernelName),
+                createPromiseFromCancellation({
+                    defaultValue: undefined,
+                    cancelAction: 'resolve',
+                    token: token
+                })
+            ];
+
+            const result = await Promise.race(kernelSearches);
+            return result ? result : this.getDefaultKernelSpec(resource);
+        }
+
+        return this.getDefaultKernelSpec(resource);
+    }
+
+    private async getKernelSpecFromActiveInterpreter(
+        resource: Resource,
+        kernelName?: string
+    ): Promise<IJupyterKernelSpec | undefined> {
+        this.activeInterpreter = await this.interpreterService.getActiveInterpreter(resource);
+
+        if (this.activeInterpreter) {
+            return this.getKernelSpec(path.join(this.activeInterpreter.path, 'share/jupyter/kernels/'), kernelName);
+        }
+    }
+
+    private async getKernelSpec(kernelPath: string, kernelName?: string): Promise<IJupyterKernelSpec | undefined> {
+        const pathExists = await this.file.directoryExists(kernelPath);
+
+        if (pathExists) {
+            let cache: IJupyterKernelSpec[] = [];
+            try {
+                // read cache file
+                cache = JSON.parse(await this.file.readFile(path.join(EXTENSION_ROOT_DIR, 'kernelSpecCache.json')));
+            } catch {
+                traceInfo('No kernelSpec cache found.');
+            }
+            this.cachedKernelSpecs = [...cache];
+
+            const kernels = await this.file.getSubDirectories(kernelPath);
+            // If no kernel name is included, return the first kernel
+            if (!kernelName) {
+                try {
+                    const kernelSpec: IJupyterKernelSpec = JSON.parse(
+                        await this.file.readFile(path.join(kernelPath, kernels[0], '/kernel.json'))
```

Comment:
```suggestion
                        await this.file.readFile(path.join(kernelPath, kernels[0], 'kernel.json'))
```

`/` is not required, should be avoided to ensure it works across platforms